browser/ui/webui/ai_chat/ai_chat_ui_page_handler.cc

```diff
@@ -53,12 +69,111 @@ using mojom::CharacterType;
+    DCHECK(!pdf_to_bitmap_converter_.is_bound());
+    GetPrintingService()->BindPdfToBitmapConverter(
+        pdf_to_bitmap_converter_.BindNewPipeAndPassReceiver());
+    pdf_to_bitmap_converter_.set_disconnect_handler(
+        base::BindOnce(&PreviewPageTextExtractor::BitmapConverterDisconnected,
+                       base::Unretained(this)));
+    if (pdf_use_skia_renderer_enabled.has_value()) {
+      pdf_to_bitmap_converter_->SetUseSkiaRendererPolicy(
+          pdf_use_skia_renderer_enabled.value());
+    }
+  }
+
+  void StartExtract() {
+    pdf_to_bitmap_converter_->GetPdfPageCount(
+        pdf_region_.Duplicate(),
+        base::BindOnce(&PreviewPageTextExtractor::OnGetPageCount,
+                       base::Unretained(this)));
+  }
+
+  void OnGetPageCount(std::optional<uint32_t> page_count) {
+    if (!page_count.has_value() || !page_count.value()) {
+      std::move(callback_).Run("");
+      return;
+    }
+    total_page_count_ = page_count.value();
+    pdf_to_bitmap_converter_->GetBitmap(
+        pdf_region_.Duplicate(), current_page_index_,
+        base::BindOnce(&PreviewPageTextExtractor::OnGetBitmap,
+                       base::Unretained(this)));
+  }
+
+  void OnGetBitmap(const SkBitmap& bitmap) {
+    if (bitmap.drawsNothing()) {
+      std::move(callback_).Run("");
+      return;
+    }
+#if BUILDFLAG(ENABLE_TEXT_RECOGNITION)
+    GetOCRText(bitmap,
+               base::BindOnce(&PreviewPageTextExtractor::OnGetTextFromImage,
+                              weak_ptr_factory_.GetWeakPtr()));
+#else
+    std::move(callback_).Run("");
+#endif
+  }
+
+  void BitmapConverterDisconnected() {
+    DLOG(ERROR) << __func__;
+    if (callback_) {
+      std::move(callback_).Run(preview_text_.str());
+    }
+  }
+
+  void OnGetTextFromImage(std::string page_content) {
+    VLOG(4) << "Page index(" << current_page_index_
+            << ") content: " << page_content;
+    preview_text_ << page_content;
+    // Stop processing if we have reached the maximum number of pages or the
+    // maximum length of the content
+    if (current_page_index_ + 1 >= kMaxPreviewPages ||
+        preview_text_.str().length() >= max_page_content_length_) {
+      std::move(callback_).Run(preview_text_.str());
+      return;
+    }
+    DCHECK_GT(total_page_count_, 0u);
+    if (current_page_index_++ < total_page_count_ - 1) {
+      preview_text_ << "\n";
+      pdf_to_bitmap_converter_->GetBitmap(
+          pdf_region_.Duplicate(), current_page_index_,
+          base::BindOnce(&PreviewPageTextExtractor::OnGetBitmap,
+                         base::Unretained(this)));
+    } else {
+      std::move(callback_).Run(preview_text_.str());
+    }
+  }
+
+ private:
+  std::stringstream preview_text_;
+  size_t current_page_index_ = 0;
+  size_t total_page_count_ = 0;
+  base::ReadOnlySharedMemoryRegion pdf_region_;
+  base::OnceCallback<void(std::string)> callback_;
+  uint32_t max_page_content_length_;
```

Comment:
`const`

Reply:
fixed in 1a018f1502b72df3f1b6a6e1f044c3f8bd9bff17